### PR TITLE
Use LightTransaction in BlockchainConnection methods

### PR DESCRIPTION
**Pull request by willclarktech**
Part 1 of #1008

Definitely fighting the type system rather than working with it here to some degree. Opening this draft PR to get feedback on the approach.

### CODE REVIEW

**Comment by webmaster128** [comment]
This looks good as far as I can see.

> Definitely fighting the type system rather than working with it here to some degree. Opening this draft PR to get feedback on the approach.

hmm, can you say what exactly is causing the fighting? The only potential improvement I see in the diff is getting rid of the default generic type of `ConfirmedTransaction` and friends to make clear, this always wraps some kind of transaction that needs to ne specified.

**Comment by willclarktech** [comment]
The main issue is here: https://github.com/iov-one/iov-core/pull/1051/files#diff-3bca4f284dd667ca2096454a4f73d657R79

With some blockchains we use a scraper/logs to get light transaction data out of the blockchain, but in other cases (eg BNS) we use a different approach and get the full transaction out (including creator). I didn't find a way to express that with types which didn't massively overcomplicate things. So the current approach has this workaround for BNS where we know that everything's ok, but the type system doesn't.

**Comment by webmaster128** [comment]
For arrays, we can use TSs type inference mechanism and use

```
public async searchTx(
    query: TransactionQuery,
  ): Promise<readonly (ConfirmedTransaction<UnsignedTransaction> | FailedTransaction)[]> {
```

to implement

```
readonly searchTx: (query: TransactionQuery) => Promise<readonly (ConfirmedTransaction<LightTransaction> | FailedTransaction)[]>;
```

i.e. the BnsConnection provides strictly more than the interface asks for in a compatible manner. However, this approach does not work for Streams.

In this case I'd pull out a private `searchTxImpl` and friends that provide the full functionality and can be used internaly.

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,12 @@
   functions.
 - @iov/bcp: Add `LightTransaction` and `WithCreator` types, and
   `isLightTransaction` helper function.
+- @iov/bcp: `ConfirmedTransaction` and `SignedTransaction` now take a type
+  argument that extends `LightTransaction` (`UnsignedTransaction` is still the
+  default).
+- @iov/bcp: Helper methods `isConfirmedTransaction` and
+  `isFailedTransaction now take a type argument that extends`LightTransaction`(`UnsignedTransaction`
```
Comment (webmaster128) [inline CHANGELOG.md]: Missing ` here

File: packages/iov-bcp/src/connection.ts [hunk 2]
```diff
@@ -148,14 +148,14 @@ export interface FailedTransaction {
   readonly message?: string;
 }
 
-export function isConfirmedTransaction(
-  transaction: ConfirmedTransaction | FailedTransaction,
-): transaction is ConfirmedTransaction {
+export function isConfirmedTransaction<T extends LightTransaction = UnsignedTransaction>(
```
Comment (webmaster128) [inline packages/iov-bcp/src/connection.ts]: If the default value needed? Isn't this inferred by from the argument?

File: packages/iov-bcp/src/connection.ts [hunk 3]
```diff
@@ -148,14 +148,14 @@ export interface FailedTransaction {
...
+  transaction: ConfirmedTransaction<T> | FailedTransaction,
+): transaction is ConfirmedTransaction<T> {
   return typeof (transaction as any).transaction !== "undefined";
 }
 
-export function isFailedTransaction(
-  transaction: ConfirmedTransaction | FailedTransaction,
+export function isFailedTransaction<T extends LightTransaction = UnsignedTransaction>(
```
Comment (webmaster128) [inline packages/iov-bcp/src/connection.ts]: If the default value needed? Isn't this inferred by from the argument?

File: packages/iov-bns/src/bnsconnection.ts [hunk 4]
```diff
@@ -566,7 +573,7 @@ export class BnsConnection implements AtomicSwapConnection {
    * It returns a stream starting the array of all existing transactions
    * and then continuing with live feeds
    */
-  public liveTx(query: TransactionQuery): Stream<ConfirmedTransaction | FailedTransaction> {
+  public liveTx(query: TransactionQuery): Stream<ConfirmedTransaction<LightTransaction> | FailedTransaction> {
```
Comment (webmaster128) [inline packages/iov-bns/src/bnsconnection.ts]: Now you can deduplictae the implementation by using `this.liveTxUnsigned`; same for `listenTx` and `searchTx`

File: packages/iov-ethereum/src/ethereumconnection.ts [hunk 5]
```diff
@@ -712,7 +713,9 @@ export class EthereumConnection implements AtomicSwapConnection {
     }
   }
 
-  public listenTx(query: TransactionQuery): Stream<ConfirmedTransaction | FailedTransaction> {
+  public listenTx(
+    query: TransactionQuery,
+  ): Stream<ConfirmedTransaction<LightTransaction> | FailedTransaction> {
```
Comment (webmaster128) [inline packages/iov-ethereum/src/ethereumconnection.ts]: Great! When you now change the return type of `searchSendTransactionsByAddressOnScraper`, we get rid of this extra network query for each transaction: `const transaction = (await this.searchTransactionsById(Parse.transactionId(tx.hash)))[0];`

File: packages/iov-bns/src/bnsconnection.ts (hunk 4 above)
Comment (willclarktech) [reply packages/iov-bns/src/bnsconnection.ts]: We still have this problem with streams though, don't we?

Comment (webmaster128) [reply packages/iov-bns/src/bnsconnection.ts]: We cannot use the result 1:1, but a simple map simipar to

```ts
this.liveTxUnsigned.map((original): ConfirmedTransaction<LightTransaction> | FailedTransaction) => {
   return original;
});```

should do the job

**Pull request merged**